### ISSUE TITLE
通过FMS/AMS的SSAS脚本不能拉取SRS发布的rtmp流

### ISSUE THREAD

**Issue by smpbo** [ossrs/srs#401]
winlinvip，你好：

在我的流媒体场景中会频繁用到rtmp流的交换，之前都是FMS/AMS之间的交换，通过FMS/AMS的脚本编程很方便使用拉取/推送rtmp流，最近有开始第3方提供SRS服务器发布的rtmp直播流，就发现不能使用FMS/AMS进行拉取，而flashplayer和ffmpeg都能正常拉取播放，以下是FMS/AMS的核心拉取脚本，同样方法拉取FMS/AMS/wowza提供的流没有问题：

``` javascript
var sourceNc=new NetConnection();
sourceNc.connect("rtmp://srs-ip/live");

var sourceNs=Stream.get("restream_name");
sourceNc.call("createStream",getResultObject("createStream"));
sourceNc.call("FCSubscribe",getResultObject("FCSubscribe"),"live_name");
sourceNs.play("live_name",-1,-1,true,sourceNc);
```

拉取之后，在FMS/AMS的日志中能正常得到NetConnection.Connect.Success事件，及之后的NetStream.Play.Reset、NetStream.Play.Start、NetStream.Data.Start，但是到NetStream.Play.Start的details显示的流名称并不是live_name，而是统一的stream，NetStream.Data.Start之后就没有事件响应，应该是流是空的就开始一直等待，日志大致如下：

```
sourceNc连接信息：
level:status
code:NetConnection.Connect.Success
description:Connection succeeded
data:
objectEncoding:3
-------------------------------------
sourceNc onBWDone
-------------------------------------
level:status
code:NetStream.Publish.Start
description:abc is now published.
details:
-------------------------------------
level:status
code:NetStream.Play.Reset
description:Playing and resetting live_name.
details:live_name
-------------------------------------
createStream Result: 1
-------------------------------------
level:status
code:NetStream.Play.Start
description:Started playing stream.
details:stream
-------------------------------------
code:NetStream.Data.Start
-------------------------------------
```

之前在CDN的对接中接触过SRS，都是给SRS-CDN提供源站，而这种拉取SRS的流的反向使用，还是第一次，此问题已苦恼多日，不知是否因为某事件调用丢失导致，还是别的原因，还请赐教。

**Comment by winlinvip** [comment]
If you need to pull the stream to SRS, can you change your approach?
Firstly, can you make SRS forward to your AMS? If you can configure SRS, you can forward the stream to your AMS. For example, CDN origin SRS forwarding to your AMS.
Additionally, can you use SRS to collect and then forward to your AMS? If you cannot modify the source SRS, can you configure your own SRS to pull the specified stream from an external SRS and then forward it to your AMS?
In other words, instead of pulling the stream from AMS to SRS, it becomes pushing the stream from SRS to AMS.

`TRANS_BY_GPT3`

**Comment by smpbo** [comment]
First of all, thank you for providing the idea of deploying an SRS as a bridge, which can indeed solve the problem. Currently, I have over 30 AMS servers in my environment, which form an application cluster. Therefore, I am reluctant to deploy another system that overlaps with it. Moreover, the method of SRS pushing to AMS is not as controllable as AMS directly pulling the stream. AMS has a higher level of initiative and stronger control, making it more convenient to operate. I'm not sure if it's due to limitations of SRS that prevent AMS from pulling, or if there are other reasons. In some projects, third parties only provide stream addresses for us to pull, rather than pushing. Currently, I can only use ffmpeg to pull and then relay the stream, similar to your idea, but the controllability is indeed not very good.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
It should be because of some different messages. I have looked at your script:

This message is missing:

Make sure to maintain the markdown structure.
```
sourceNc.call("FCSubscribe",getResultObject("FCSubscribe"),"live_name");
```


Is your current issue that you need SRS to provide you with a stream name instead of a unified stream string? Why do you need to know the name of the stream you are pulling? Isn't the stream name specified by you?

`TRANS_BY_GPT3`

**Comment by smpbo** [comment]
FCSubscribe, I have tested it and it doesn't have any impact whether it is added or not. It was added by CDN's technical brother before the createStream call, and after debugging, I was able to publish the stream to SRS using AMS (which was not possible before). However, I couldn't pull and play the stream. I can only guess that the "stream" is the output stream name returned by SRS.

I used ffmpeg to pull and play the stream and analyzed its debug information. I found that when ffmpeg plays rtmp, it will connect, createStream, FCSubscribe, and _checkbw, and then it starts playing after receiving the meta. So I tried to simulate the entire process of ffmpeg, but I found no effect. Regardless of removing or changing the order or different combinations of calls, AMS cannot play the stream. It's quite strange that nothing happens after the NetStream.Data.Start event.

However, this is indeed a relatively niche use case. After all, AMS is gradually fading away from the scene. If this issue doesn't have a significant impact on the software and its usage, we can just discuss it. The only thing I feel is that SRS lacks the same compatibility as AMS in terms of stream handshake. Of course, if SRS becomes more popular in the future, this won't be a problem anymore.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
Well, SRS indeed has not made compatibility with AMS for pulling streams, and such compatibility would require debugging.
AMS is relatively closed, and when it pulls streams from others, it sends some special packets, so AMS is often used as the source station, and others need to adapt to it.
If you need to fix this issue, then I will take a look at it in SRS2 when I have time, and I will need your cooperation to set up an environment for me.
If you think you can use SRS for pulling streams and don't need to fix this problem, then you can just ignore it.

`TRANS_BY_GPT3`

**Comment by smpbo** [comment]
AMS is, after all, commercial software, hehe.
If you have the time and happen to be able to fix it conveniently, that would be great. I am more than willing to cooperate fully. I am very familiar with FMS/AMS, and in the current environment, I also want to learn more about SRS.
My email is smpbo@163.com. If it's convenient, could you provide me with an IM contact for future collaboration? Thank you for your support.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
WeChat group: https://github.com/ossrs/srs/wiki/v1_CN_Contact#wechat

`TRANS_BY_GPT3`

**Comment by limbo0312** [comment]
I have also encountered a similar problem. May I ask how you will proceed with handling or avoiding it in the future? @smpbo @winlinvip

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
From SRS to FMS, there can be multiple ways:

1. SRS serves as the Edge, and FMS serves as the Origin.
2. SRS forwards the stream to FMS.
3. Pulling the stream from SRS to FMS using FFMPEG or using SRS's Ingest feature are the same thing.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
AMS has conducted research before, and when AMS is used for Edge, it will use a proprietary protocol instead of the standard RTMP protocol, so SRS does not support AMS as a source.

`TRANS_BY_GPT3`
